src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -381,6 +383,67 @@ private bool StartCore(ProcessStartInfo startInfo)
             return true;
         }
 
+        private bool ForkAndExecProcess(
+            string filename, string[] argv, string[] envp, string cwd,
+            bool redirectStdin, bool redirectStdout, bool redirectStderr,
+            bool setCredentials, uint userId, uint groupId,
+            out int stdinFd, out int stdoutFd, out int stderrFd,
+            bool throwOnNoExec = true)
+        {
+            if (string.IsNullOrEmpty(filename))
+            {
+                throw new Win32Exception(Interop.Error.ENOENT.Info().RawErrno);
+            }
+
+            // Lock to avoid races with OnSigChild
+            // By using a ReaderWriterLock we allow multiple processes to start concurrently.
+            s_processStartLock.EnterReadLock();
+            try
+            {
+                int childPid;
+
+                // Invoke the shim fork/execve routine.  It will create pipes for all requested
+                // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
+                // descriptors, and execve to execute the requested process.  The shim implementation
+                // is used to fork/execve as executing managed code in a forked process is not safe (only
+                // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
+                int errno = Interop.Sys.ForkAndExecProcess(
+                    filename, argv, envp, cwd,
+                    redirectStdin, redirectStdout, redirectStderr,
+                    setCredentials, userId, groupId,
+                    out childPid,
+                    out stdinFd, out stdoutFd, out stderrFd);
+
+                if (errno == 0)
+                {
+                    // Ensure we'll reap this process.
+                    // note: SetProcessId will set this if we don't set it first.
+                    _waitStateHolder = new ProcessWaitState.Holder(childPid, isNewChild: true);
+
+                    // Store the child's information into this Process object.
+                    Debug.Assert(childPid >= 0);
+                    SetProcessId(childPid);
+                    SetProcessHandle(new SafeProcessHandle(childPid));
+
+                    return true;
+                }
+                else
+                {
+                    if ((throwOnNoExec == false)
+                        && (Interop.Sys.ConvertErrorPlatformToPal(errno) == Interop.Error.ENOEXEC))
```

Comment:
Rather than using ConvertErrorPlatformToPal directly, you can use ErrorInfo, e.g. `new ErrorInfo(Interop.Error.ENOEXEC).RawErrno == errno` or `new ErrorInfo(errno).Error == Interop.Error.ENOEXEC`.  We generally don't directly use ConvertErrorPlatformToPal/PalToPlatform, instead preferring to use ErrorInfo throughout the code; they're really only meant to be used by ErrorInfo.